### DISCUSSION TITLE
Check for errors on fields created with `useFieldArray`

### DISCUSSION THREAD

**Discussion by lukebennett88** [react-hook-form/react-hook-form#9331] [Q&A]
First of all, thanks for the great library @bluebill1049 and team, I really enjoy using it and appreciate all the hard work involved.

I have a few questions, I'll do my best to keep this as easy to understand as possible, but apologies if that doesn't end up being the case 😬

I'm working on an app that lets users add multiple items to a list, this seemed like the perfect use-case for `useFieldArray`, which I haven't used before now.

I'm using [Zod](https://zod.dev/) and the [zodResolver](https://github.com/react-hook-form/resolvers) for schema validation.

The first issue I ran into is that `useFieldArray` wants me to provide default values when I use the `append` method. For the schema I'm using, I want all fields to be required, but I don't want to provide default values as there are no reasonable defaults. Is there are way to do this without casting or using `@ts-ignore`?

The second, and more important question is how to check for errors.
When using the regular `react-hook-form` API (e.g. `register` without `useFieldArray`) I'd use a utility like this:

```tsx
import { type FieldValues, type UseFormReturn } from "react-hook-form";

export function validateField<TFieldValues extends FieldValues>(
	form: UseFormReturn<TFieldValues>,
	key: keyof TFieldValues
) {
	const error = form.formState.errors[key];
	if (error) {
		return {
			message: error.message,
			tone: "critical" as const,
		};
	}
}

/**
 * In this example, `Field` takes a `tone` prop for styling, and a `message` 
 * prop we can use for our error.
 */
<Field label="Description" {...validateField(form, "description")}>
	<TextInput
		{...form.register("description")}
	>
		
	</TextInput>
</Field>
```

This might be more of a TypeScript question, but is there a recommended way to check for errors for a field in `useFieldArray`?

**Comment by bluebill1049** [comment]
hey @lukebennett88 

> The first issue I ran into is that useFieldArray wants me to provide default values when I use the append method. For the schema I'm using, I want all fields to be required, but I don't want to provide default values as there are no reasonable defaults. Is there are way to do this without casting or using @ts-ignore?

Each inputs should have a default value, It could be just an empty string. This is required as each input with `useFieldArray` is uncontrolled by nature which will required to render the input with a defaultValue.

something as simple as `<input />` actually has a default value of empty string as well.

> When using the regular react-hook-form API (e.g. register without useFieldArray) I'd use a utility like this:

I would prefer not to have a utility function to get error, just leverage the type from the actual errors.

Such as `errors.fieldArray?.[0]?.firstName?.message`.

Hope the above make more sense, feel free to follow up with more questions.

**Reply by lukebennett88** [reply]
> Each inputs should have a default value, It could be just an empty string. This is required as each input with `useFieldArray` is uncontrolled by nature which will required to render the input with a defaultValue.

The main problem I have is with numbers. I don't really want to value to be pre-filled with `0`


> I would prefer not to have a utility function to get error, just leverage the type from the actual errors.

I wouldn't expect the library to offer a utility for this, but I'd personally find it very useful. Otherwise there is a lot of extremely similar code that is also quite hard to read. e.g. 

```tsx
<Field
	{...(form.formState.errors.lots?.[index]?.reservePriceCents?.message
		? {
				message:
					form.formState.errors.lots?.[index]?.reservePriceCents?.message,
				tone: "critical",
			}
		: {})}
>
	<TextInput
		message={
			form.formState.errors.lots?.[index]?.reservePriceCents?.message
		}
		{...form.register(`lots.${index}.reservePriceCents`, {
			valueAsNumber: true,
		})}
	>
		Reserve price
	</TextInput>
</Field>
```

**Reply by bluebill1049** [reply]
> The main problem I have is with numbers. I don't really want to value to be pre-filled with 0

if you are referring to input `type="number"` then `defaultValue` is `""` native input can only use valueAsNumber to retrieve the number type, if you are using zod, then the transform is done there.

> I wouldn't expect the library to offer a utility for this, but I'd personally find it very useful. Otherwise there is a lot of extremely similar code that is also quite hard to read. e.g.

I got it, it is probably losing some of the type checks there, but you can always build your function to take a generic to type guard the error message.

**Reply by lukebennett88** [reply]
> if you are referring to input `type="number"` then `defaultValue` is `""` native input can only use valueAsNumber to retrieve the number type, if you are using zod, then the transform is done there.

I'm using a regular `type="text"` input with `valueAsNumber: true` (number inputs have some a11y issues so I tend to avoid them). I'm not sure how I'd provide an empty string as the default value without TS yelling at me, but I guess that's more of a Zod issue that a `react-hook-form` issue 😅
